tgext/utils/csrf.py

```diff
@@ -20,38 +54,47 @@ def _get_conf():
     """
     conf = tg.config
     csrf_secret = conf['csrf.secret']
-    csrf_token_name = str(conf.get('csrf.token_name', CSRF_TOKEN))
-    csrf_path = conf.get('csrf.path', ROOT).encode(ENCODING)
-    try:
-        cookie_expires = int(conf.get('csrf.expires', EXPIRES))
-    except ValueError:
-        cookie_expires = EXPIRES
-    return csrf_secret, csrf_token_name, csrf_path, cookie_expires
+    csrf_token_name = conf['csrf.token_name']
+    csrf_path = conf['csrf.path'].encode(ENCODING)
+    cookie_expires = conf['csrf.expires']
+    handler = conf['csrf.error_handler']
+    return csrf_secret, csrf_token_name, csrf_path, cookie_expires, handler
 
 
 def _generate_csrf_token():
     """
     Generate and set new CSRF token in cookie. The generated token is set to
     ``request.csrf_token`` attribute for easier access by other functions.
     """
-    secret, token_name, path, expires = _get_conf()
-    sha256 = hashlib.sha256()
-    sha256.update(os.urandom(8))
-    token = sha256.hexdigest().encode(ENCODING)
-    tg.response.signed_cookie(token_name, token, secret,
+    secret, token_name, path, expires, _ = _get_conf()
+    session_id = tg.session['_id']
```

Comment:
This now means we must document the need to have sessions enabled. 
PS: Not sure I get the need to use the `session_id`, I guess the purpose is to make the token unique per user, but the session_id is a value that still comes from the client, so can be easily crafted. Maybe we meant to use the `user_id` which is instead unknown to the client?

Reply:
I'm not sure the session_id comes from the client, it is generated here https://github.com/bbangert/beaker/blob/889d3055a4ca31b55a0b0681b00f2973b3250d88/beaker/session.py#L35 when the session is new or regenerated

Reply:
Sorry for not being specific. It's true that we generate a value and send it back to the client. So in theory you could see the session id as something that was generated by the server.

But in reality we just trust that value, you can send back _any_ session_id. If it doesn't exist it will just be an empty session. So in theory anyone can mimic that session id. I think that the only way that session id can be reliable is if the session framework was configured to use a signed cookie, which is not enforced nor guaranteed to be what users had done.